core/src/validator.rs

```diff
@@ -2833,6 +2798,121 @@ pub fn is_snapshot_config_valid(snapshot_config: &SnapshotConfig) -> bool {
     }
 }
 
+/// [`RuntimeRegistry`] structure holds runtimes for different components of the
+/// validator.
+///
+/// This structure might be used when the validator is started inside the tokio
+/// runtime context or outside. Test-validator crate and tests may start the
+/// validator in a tokio runtime context which forces us to use the same runtime
+/// because a nested runtime will cause panic at drop. Outside test-validator
+/// crate, we always need a tokio runtime (and the respective handle) to
+/// initialize the turbine QUIC endpoint.
+struct RuntimeRegistry {
+    runtimes: RuntimeOption,
+}
+
+/// [`RuntimeOption`] represents the current runtime if available or a dedicated
+/// runtime.
+enum RuntimeOption {
+    Current,
+    Dedicated(HashMap<RuntimeId, TokioRuntime>),
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+enum RuntimeId {
+    TpuClientNext,
+    TurbineQuicEndpoint,
+    RepairQuicEndpoints,
+}
+
+impl fmt::Display for RuntimeId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            RuntimeId::TpuClientNext => "solTpuClientRt",
+            RuntimeId::TurbineQuicEndpoint => "solTurbineQuic",
+            RuntimeId::RepairQuicEndpoints => "solRepairQuic",
+        };
+        f.write_str(s)
+    }
+}
+
+impl RuntimeRegistry {
+    fn new() -> Self {
+        match tokio::runtime::Handle::try_current() {
+            Ok(_) => Self {
+                runtimes: RuntimeOption::Current,
+            },
+            Err(_) => Self {
+                runtimes: RuntimeOption::Dedicated(HashMap::new()),
+            },
+        }
+    }
+
+    /// Create a new runtime with the given name and number of worker threads
+    /// and insert it into the registry.
+    ///
+    /// If a runtime with the same name already exists, this function will
+    /// panic.
+    fn insert_runtime(&mut self, id: RuntimeId, num_threads: usize) {
+        if let RuntimeOption::Dedicated(runtimes) = &mut self.runtimes {
+            if !runtimes.contains_key(&id) {
```

Comment:
*minor*

It is probably better to reduce the indentation:

```suggestion
    fn insert_runtime(&mut self, id: RuntimeId, num_threads: usize) {
        let RuntimeOption::Dedicated(runtimes) = &mut self.runtimes else {
            /* Will be using a thread local runtime. */
            return;
        }

        let Entry::Vacant(entry) = runtimes.entry(&id) else {
            /* This runtime was already configured. */
            return;
        }

        entry.insert(
            tokio::runtime::Builder::new_multi_thread()
                .enable_all()
                .worker_threads(num_threads)
                .thread_name(id.to_string())
                .build()
                .unwrap()
        );
```